### PR TITLE
Make id optional

### PR DESCRIPTION
**Pull request by RoachMech**
Json response without an id is now possible. Simply return false on the getId method of a Resource's Serializer and no id will be returned.

Documents without an id are conform the JSON API v1.0 specification. For example, see: http://jsonapi.org/format/#crud-creating

### CODE REVIEW

**Comment by byCedric** [comment]
I'm curious why you have used a `false` return value from the `getId` methods of both the serializer and the resource. Why haven't you simply used a `null` value or `''` (empty string) for example, since they are both _falsy_ values.

**Comment by byCedric** [comment]
My curiosity is mostly sparked from the rule that both `id` and `type` _Resource_ properties [*MUST* be strings](http://jsonapi.org/format/#document-resource-object-identification). Therefore it would make more sense to me to return an empty string from the `getId` methods.

If you only use strings to as return value, you can replace the line [#166](https://github.com/tobscure/json-api/pull/100/commits/fae88ce6768aa5f22257bc3c24f9a697c49f6404#diff-a397cefd933eb7520f6375a22099874cR166) with
`return $this->serializer->getId($this->data);` making it cleaner.

**Comment by RoachMech** [comment]
Hi Cedric,

Thanks for your feedback! I've changed the return value on the getId() method as you suggested.

**Comment by byCedric** [comment]
You are welcome! When I get home today I will check on why the tests are failing. That will be around 19:00 (CEST). Maybe someone with admin access can check on the StyleCI errors?

**Comment by byCedric** [comment]
Well I dug a little bit deeper into the issue, and here is my contribution. This pull request tries to tackle the following exception of the JSON API spec v1.0.

> The `id` member is not required when the resource object _originates at the client_ and represents a new resource to be created on the server.

This says that when trying to create a new resource instance, the client can create a resource that omits the `id` property. Mainly because it is not available and it doesn't make sense to use `{ "id": null }`. I think actually don't think this library was designed to be used as _client_. Mainly because in the [README](https://github.com/tobscure/json-api/blob/master/README.md) the word _responses_ is used (`JSON-API responses in PHP.`), implicating this library should be used as _server_. So I think this feature should be discussed first. 

Besides the point if this should be implemented or not, this pull request needs a little more work. It needs some proper testing and it contains changes that has implications on other files and tests. E.g. the serializer `getId` method should be able to return `null` for this feature. For the full list of things that should be done when contributing, see the [contributing guidelines](https://github.com/tobscure/json-api/blob/master/CONTRIBUTING.md).

**TL;DR;**
- Try to start a discussion in a new issue. We can openly discus there if this feature should be implemented and/or how.
- Try to adhere the [contributing guidelines](https://github.com/tobscure/json-api/blob/master/CONTRIBUTING.md) by adding tests for next pull requests.

**Comment by RoachMech** [comment]
I completely understand your opinion. I will openly ask whether this library should allow omitting the `id` property and thus can be used as a _client_.

**Comment by franzliedke** [comment]
Howdy, interesting suggestion.

However, I mostly agree with @byCedric that this is not needed for the current feature set of the library. Can you explain what exactly the use case is? Are you using this to build a JSON-API client or what is going on?

**Comment by tobscure** [comment]
Closing for now in favour of discussion in #101.
